<a href="https://colab.research.google.com/github/anabiarochar/Mestrado_AlgoritmosProgramacao/blob/main/Projeto1/4_MergeSort.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Projeto 1 - Algoritmo de Ordenação Inserção (Insertion Sort)

Abaixo estão os algorítmos para ordenação, utilizando o método inserção de ordenação, com entradas de vetores de tamanhos 1.000, 10.000 e 100.000

- aleatórias
- ordenadas em forma crescente
- ordenadas em forma decrescente  

Abaixo de cada algorítmo estão os tempos observados somando 4 tentativas.

##Vetor em ordem aleatória com 1000 elementos:

In [4]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_1000_aleatorio.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método Merge Sort
def mergeSort(arr):
    n = len(arr) # n=len(V)
    if n <= 1:
        return arr

    meio = n // 2

    esquerda = []
    for i in range(meio):
        esquerda.append(arr[i])

    direita = []
    for i in range(meio, n):
        direita.append(arr[i])

    esquerda = mergeSort(esquerda)
    direita = mergeSort(direita)

    return merge(esquerda, direita)

def merge(esquerda, direita):
    resultado = []
    i = j = 0

    while i < len(esquerda) and j < len(direita):
        if esquerda[i] < direita[j]:
            resultado.append(esquerda[i])
            i += 1
        else:
            resultado.append(direita[j])
            j += 1

    resultado.extend(esquerda[i:])
    resultado.extend(direita[j:])
    return resultado

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = mergeSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.002539 segundos
[1, 3, 8, 9, 9, 10, 10, 10, 10, 12]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.002458 segundos  
Tentativa 2 - Tempo de execução: 0.002509 segundos  
Tentativa 3 - Tempo de execução: 0.004544 segundos  
Tentativa 4 - Tempo de execução: 0.002539 segundos  

##Vetor em ordem aleatória com 10000 elementos:

In [8]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_10000_aleatorio.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método Merge Sort
def mergeSort(arr):
    n = len(arr) # n=len(V)
    if n <= 1:
        return arr

    meio = n // 2

    esquerda = []
    for i in range(meio):
        esquerda.append(arr[i])

    direita = []
    for i in range(meio, n):
        direita.append(arr[i])

    esquerda = mergeSort(esquerda)
    direita = mergeSort(direita)

    return merge(esquerda, direita)

def merge(esquerda, direita):
    resultado = []
    i = j = 0

    while i < len(esquerda) and j < len(direita):
        if esquerda[i] < direita[j]:
            resultado.append(esquerda[i])
            i += 1
        else:
            resultado.append(direita[j])
            j += 1

    resultado.extend(esquerda[i:])
    resultado.extend(direita[j:])
    return resultado

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = mergeSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.032530 segundos
[3, 3, 4, 8, 9, 9, 9, 11, 12, 12]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.032408 segundos  
Tentativa 2 - Tempo de execução: 0.031459 segundos  
Tentativa 3 - Tempo de execução: 0.032583 segundos  
Tentativa 4 - Tempo de execução: 0.032530 segundos  

##Vetor em ordem aleatória com 100000 elementos:

In [12]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_100000_aleatorio.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método Merge Sort
def mergeSort(arr):
    n = len(arr) # n=len(V)
    if n <= 1:
        return arr

    meio = n // 2

    esquerda = []
    for i in range(meio):
        esquerda.append(arr[i])

    direita = []
    for i in range(meio, n):
        direita.append(arr[i])

    esquerda = mergeSort(esquerda)
    direita = mergeSort(direita)

    return merge(esquerda, direita)

def merge(esquerda, direita):
    resultado = []
    i = j = 0

    while i < len(esquerda) and j < len(direita):
        if esquerda[i] < direita[j]:
            resultado.append(esquerda[i])
            i += 1
        else:
            resultado.append(direita[j])
            j += 1

    resultado.extend(esquerda[i:])
    resultado.extend(direita[j:])
    return resultado

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = mergeSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.425543 segundos
[2, 5, 6, 6, 8, 9, 10, 13, 14, 14]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.420817 segundos  
Tentativa 2 - Tempo de execução: 0.764584 segundos  
Tentativa 3 - Tempo de execução: 0.436991 segundos  
Tentativa 4 - Tempo de execução: 0.425543 segundos  

##Vetor em ordem crescente com 1000 elementos:

In [16]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_1000_crescente.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método Merge Sort
def mergeSort(arr):
    n = len(arr) # n=len(V)
    if n <= 1:
        return arr

    meio = n // 2

    esquerda = []
    for i in range(meio):
        esquerda.append(arr[i])

    direita = []
    for i in range(meio, n):
        direita.append(arr[i])

    esquerda = mergeSort(esquerda)
    direita = mergeSort(direita)

    return merge(esquerda, direita)

def merge(esquerda, direita):
    resultado = []
    i = j = 0

    while i < len(esquerda) and j < len(direita):
        if esquerda[i] < direita[j]:
            resultado.append(esquerda[i])
            i += 1
        else:
            resultado.append(direita[j])
            j += 1

    resultado.extend(esquerda[i:])
    resultado.extend(direita[j:])
    return resultado

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = mergeSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.001879 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.006217 segundos  
Tentativa 2 - Tempo de execução: 0.001821 segundos  
Tentativa 3 - Tempo de execução: 0.001828 segundos  
Tentativa 4 - Tempo de execução: 0.001879 segundos

##Vetor em ordem crescente com 10000 elementos:

In [20]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_10000_crescente.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método Merge Sort
def mergeSort(arr):
    n = len(arr) # n=len(V)
    if n <= 1:
        return arr

    meio = n // 2

    esquerda = []
    for i in range(meio):
        esquerda.append(arr[i])

    direita = []
    for i in range(meio, n):
        direita.append(arr[i])

    esquerda = mergeSort(esquerda)
    direita = mergeSort(direita)

    return merge(esquerda, direita)

def merge(esquerda, direita):
    resultado = []
    i = j = 0

    while i < len(esquerda) and j < len(direita):
        if esquerda[i] < direita[j]:
            resultado.append(esquerda[i])
            i += 1
        else:
            resultado.append(direita[j])
            j += 1

    resultado.extend(esquerda[i:])
    resultado.extend(direita[j:])
    return resultado

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = mergeSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.042734 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.022288 segundos  
Tentativa 2 - Tempo de execução: 0.024113 segundos  
Tentativa 3 - Tempo de execução: 0.022821 segundos  
Tentativa 4 - Tempo de execução: 0.042734 segundos  

##Vetor em ordem crescente com 100000 elementos:

In [24]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_100000_crescente.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método Merge Sort
def mergeSort(arr):
    n = len(arr) # n=len(V)
    if n <= 1:
        return arr

    meio = n // 2

    esquerda = []
    for i in range(meio):
        esquerda.append(arr[i])

    direita = []
    for i in range(meio, n):
        direita.append(arr[i])

    esquerda = mergeSort(esquerda)
    direita = mergeSort(direita)

    return merge(esquerda, direita)

def merge(esquerda, direita):
    resultado = []
    i = j = 0

    while i < len(esquerda) and j < len(direita):
        if esquerda[i] < direita[j]:
            resultado.append(esquerda[i])
            i += 1
        else:
            resultado.append(direita[j])
            j += 1

    resultado.extend(esquerda[i:])
    resultado.extend(direita[j:])
    return resultado

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = mergeSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.275732 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.271876 segundos  
Tentativa 2 - Tempo de execução: 0.269218 segundos  
Tentativa 3 - Tempo de execução: 0.270475 segundos  
Tentativa 4 - Tempo de execução: 0.275732 segundos  

##Vetor em ordem decrescente com 1000 elementos:

In [28]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_1000_decrescente.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método Merge Sort
def mergeSort(arr):
    n = len(arr) # n=len(V)
    if n <= 1:
        return arr

    meio = n // 2

    esquerda = []
    for i in range(meio):
        esquerda.append(arr[i])

    direita = []
    for i in range(meio, n):
        direita.append(arr[i])

    esquerda = mergeSort(esquerda)
    direita = mergeSort(direita)

    return merge(esquerda, direita)

def merge(esquerda, direita):
    resultado = []
    i = j = 0

    while i < len(esquerda) and j < len(direita):
        if esquerda[i] < direita[j]:
            resultado.append(esquerda[i])
            i += 1
        else:
            resultado.append(direita[j])
            j += 1

    resultado.extend(esquerda[i:])
    resultado.extend(direita[j:])
    return resultado

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = mergeSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.003225 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.003413 segundos  
Tentativa 2 - Tempo de execução: 0.001863 segundos  
Tentativa 3 - Tempo de execução: 0.002033 segundos  
Tentativa 4 - Tempo de execução: 0.003225 segundos  

##Vetor em ordem decrescente com 10000 elementos:

In [32]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_10000_decrescente.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método Merge Sort
def mergeSort(arr):
    n = len(arr) # n=len(V)
    if n <= 1:
        return arr

    meio = n // 2

    esquerda = []
    for i in range(meio):
        esquerda.append(arr[i])

    direita = []
    for i in range(meio, n):
        direita.append(arr[i])

    esquerda = mergeSort(esquerda)
    direita = mergeSort(direita)

    return merge(esquerda, direita)

def merge(esquerda, direita):
    resultado = []
    i = j = 0

    while i < len(esquerda) and j < len(direita):
        if esquerda[i] < direita[j]:
            resultado.append(esquerda[i])
            i += 1
        else:
            resultado.append(direita[j])
            j += 1

    resultado.extend(esquerda[i:])
    resultado.extend(direita[j:])
    return resultado

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = mergeSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.022781 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.022821 segundos  
Tentativa 2 - Tempo de execução: 0.039824 segundos  
Tentativa 3 - Tempo de execução: 0.022876 segundos  
Tentativa 4 - Tempo de execução: 0.022781 segundos  

##Vetor em ordem decrescente com 100000 elementos:

In [36]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_100000_decrescente.txt"

vetor = []

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método Merge Sort
def mergeSort(arr):
    n = len(arr) # n=len(V)
    if n <= 1:
        return arr

    meio = n // 2

    esquerda = []
    for i in range(meio):
        esquerda.append(arr[i])

    direita = []
    for i in range(meio, n):
        direita.append(arr[i])

    esquerda = mergeSort(esquerda)
    direita = mergeSort(direita)

    return merge(esquerda, direita)

def merge(esquerda, direita):
    resultado = []
    i = j = 0

    while i < len(esquerda) and j < len(direita):
        if esquerda[i] < direita[j]:
            resultado.append(esquerda[i])
            i += 1
        else:
            resultado.append(direita[j])
            j += 1

    resultado.extend(esquerda[i:])
    resultado.extend(direita[j:])
    return resultado

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = mergeSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.306561 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.294136 segundos  
Tentativa 2 - Tempo de execução: 0.287729 segundos  
Tentativa 3 - Tempo de execução: 0.285567 segundos  
Tentativa 4 - Tempo de execução: 0.306561 segundos  

##Resumo dos Tempos de Execução

| Caso do vetor | n | Tentativa 1 | Tentativa 2 | Tentativa 3 | Tentativa 4 | Média |
|---|---:|---:|---:|---:|---:|---:|
| Aleatório | 1.000 | 0.002458 | 0.002509 | 0.004544 | 0.002539 | 0.003013 |
| Aleatório | 10.000 | 0.032408 | 0.031459 | 0.032583 | 0.032530 | 0.032245 |
| Aleatório | 100.000 | 0.420817 | 0.764584 | 0.436991 | 0.425543 | 0.511984 |
| Crescente | 1.000 | 0.006217 | 0.001821 | 0.001828 | 0.001879 | 0.002936 |
| Crescente | 10.000 | 0.022288 | 0.024113 | 0.022821 | 0.042734 | 0.027989 |
| Crescente | 100.000 | 0.271876 | 0.269218 | 0.270475 | 0.275732 | 0.271825 |
| Decrescente | 1.000 | 0.003413 | 0.001863 | 0.002033 | 0.003225 | 0.002634 |
| Decrescente | 10.000 | 0.022821 | 0.039824 | 0.022876 | 0.022781 | 0.027076 |
| Decrescente | 100.000 | 0.294136 | 0.287729 | 0.285567 | 0.306561 | 0.293498 |